--- android_icu4j/src/main/java/android/icu/lang/UCharacter.java	2017-10-13 00:22:56.775382358 -0700
+++ android_icu4j/src/main/java/android/icu/lang/UCharacter.java	2017-10-13 00:23:10.887535857 -0700
@@ -53,16 +53,6 @@
  * more convenient in Java to have a separate primitive datatype for them,
  * ints suffice in the meantime.
  *
- * <p>To use this class please add the jar file name icu4j.jar to the
- * class path, since it contains data files which supply the information used
- * by this file.<br>
- * E.g. In Windows <br>
- * <code>set CLASSPATH=%CLASSPATH%;$JAR_FILE_PATH/ucharacter.jar</code>.<br>
- * Otherwise, another method would be to copy the files uprops.dat and
- * unames.icu from the icu4j source subdirectory
- * <i>$ICU4J_SRC/src/android.icu.impl.data</i> to your class directory
- * <i>$ICU4J_CLASS/android.icu.impl.data</i>.
- *
  * <p>Aside from the additions for UTF-16 support, and the updated Unicode
  * properties, the main differences between UCharacter and Character are:
  * <ul>
@@ -84,11 +74,6 @@
  *      as having numeric values.  This is a semantic change from ICU4J 1.3.1.
  * </ul>
  * <p>
- * Further detail on differences can be determined using the program
- *        <a href=
- * "http://source.icu-project.org/repos/icu/icu4j/trunk/src/com/ibm/icu/dev/test/lang/UCharacterCompare.java">
- *        android.icu.dev.test.lang.UCharacterCompare</a>
- * <p>
  * In addition to Java compatibility functions, which calculate derived properties,
  * this API provides low-level access to the Unicode Character Database.
  * <p>
