--- android_icu4j/src/main/java/android/icu/util/IslamicCalendar.java	2017-08-25 08:53:44.225866216 +0100
+++ android_icu4j/src/main/java/android/icu/util/IslamicCalendar.java	2017-08-25 08:54:12.453921200 +0100
@@ -67,12 +67,7 @@
  * Like the Islamic religious calendar, <em>Umm al-Qura</em> is also based
  * on the sighting method of the crescent moon but is standardized by Saudi Arabia.
  * <p>
- * The {@link #setCalculationType(CalculationType) setCalculationType} method determines
- * which approach is used to determine the start of a month.  By default, the
- * fixed-cycle <em>civil</em> calendar is used.  However, if <code>setCalculationType(ISLAMIC)</code>
- * is called, an approximation of the true lunar calendar will be used.
- * Similarly, if <code>setCalculationType(ISLAMIC_UMALQURA)</code> is called, an approximation
- * of the Umm al-Qura lunar calendar will be used.
+ * The fixed-cycle <em>civil</em> calendar is used.
  * <p>
  * This class should not be subclassed.</p>
  * <p>
